src/components/AlchemyRow.vue

```diff
@@ -0,0 +1,185 @@
+      type="number" :min="alchemy.upgrades[group][props.idx]"
+      :value="alchemy.goals[group][props.idx]"
+      @change="handleAlchemyGoal($event, group, props.idx)"/>
+  </td>
+  <td class="centered columnSpan-2 effect"> 
+    {{effectChange()}}
+  </td>
+  <table class="filled">
+    <tr :style="{display: 'flex'}">
+      <td
+        :style="{width: '25%', whiteSpace: 'pre', marginRight: '5%'}" 
+        v-for="(thing, idx) in computeMaterials()"
+        :key="idx"
+      >
+        <th>{{thing[0]}}</th>
+        <tr :style="{display: 'inline-block', textAlign: 'right', width: '100%'}">
+          {{thing[1]}}
+          </tr>
+      </td>
+    </tr>
+  </table> 
+</template>
+
+
+<script lang="ts">
+import { computed, defineComponent,} from "vue";
+import { useState} from "~/State";
+import GameAsset from "~/components/GameAsset.vue";
+import { Assets, Growth } from "~/composables/Utilities";
+import { Color , Alch, Bubble} from "~/composables/Alchemy";
+import bubblesData from "~/data/bubbles.json"
+
+
+
+export default defineComponent({
+  name: "AlchemyRow",
+  components: {
+    GameAsset,
+  },
+  props: {
+    group: {
+      required: true,
+      type: String,
+    },
+    discount: {
+      required: true,
+      type: Array,
+    },
+    idx: {
+      required: true, 
+      type: Number,
+    }
+  },
+  methods: {
+    customChange (event) {
+      console.log("Handle leveling bubble input")
+      console.log("Only emit when id == 14 or (id == 6 and group == yellow)")
+      this.alchemy.upgrades[this.props.group][this.props.idx] = parseInt(event.target.value);
+      if (this.props.idx == 14 || (this.props.idx == 6 && this.props.group == "Yellow")) {
+        this.$emit("custom-change", event.target.value);
+        console.log(`Emitted [${15-event.target.id}] ${event.target.value}`);
+      }
+    }
+  },
+  setup(props) {
+    const state = useState();
+
+    const alchemy = computed({
+      get: () => state.value.alchemy,
+      set: (value) => (state.value.alchemy = value),
+    });
+   
+    //  TODO: I think this should be done in State.ts
+    //    Possibly also needed for goals.
+    // if (alchemy.value.upgrades[props.group] !== undefined) {
+    //   let upgradeDiff = amountBubbles - alchemy.value.upgrades[props.group].length;
+    //   for (let i = 0; i < upgradeDiff; i += 1) {
+    //     console.log(`Adding values in ${props.group} ${upgradeDiff}`)
+    //     alchemy.value.upgrades[props.group].push(0);
+    //   }
+    // } 
+
+    // TODO:  This could maybe move to the customChange method
+    const handleAlchemyGoal = (ev: Event, color: Color, i: number) => {
+      let target = <HTMLInputElement>ev.target;
+      let val = target ? target.value : "0";
+      alchemy.value.goals[color][i] = parseInt(val);
+    };
+    
+
+    // Compute all materials needed for the wanted upgrades
+    const computeMaterials = () =>  {
+    
+      let bubble = bubblesData[props.group][props.idx];
```

Comment:
As the data `bubblesData.json` is always accessed as `[Color][number]`, but is actually written in the form of `[Color][String]`, type inference breaks. As all the props of the inner object are just number strings, everything still works.
By just changing the inner object to an array, the type inference would be solved.
As an example:
```
{
  "Orange": {
    "0": {...},
    ...
  },...
}
```
changing to

```
{
  "Orange": [
    {...},
    ...
  ],...
}
```
Would solve the problem

Reply:
This has been applied.